release/nightly_tests/dataset/multi_node_train_benchmark.py

```diff
@@ -145,43 +185,210 @@ def crop_and_flip_image(row):
+        it = train.get_dataset_shard(f"train_{worker_rank}")
+    else:
+        it = train.get_dataset_shard("train")
     device = train.torch.get_device()
 
+    batch_iter = None
+    if args.use_torch:
+        torch_num_workers = args.torch_num_workers
+        if torch_num_workers is None:
+            torch_num_workers = 256
+        batch_iter = get_torch_data_loader(
+            worker_rank=worker_rank,
+            batch_size=args.batch_size,
+            num_workers=torch_num_workers,
+            transform=get_transform(True),
+        )
+    elif args.use_mosaic:
+        target_epoch_size = get_mosaic_epoch_size(
+            args.num_workers, target_worker_gb=args.target_worker_gb
+        )
+        print(
+            "Epoch size:",
+            target_epoch_size if target_epoch_size is not None else "all",
+            "images",
+        )
+
+        num_physical_nodes = (
+            ray.train.get_context().get_world_size()
+            // ray.train.get_context().get_local_world_size()
+        )
+        torch_num_workers = args.torch_num_workers
+        if torch_num_workers is None:
+            torch_num_workers = os.cpu_count()
+        # Divide by the number of Train workers because each has its own dataloader.
+        torch_num_workers //= ray.train.get_context().get_local_world_size()
+
+        batch_iter = get_mosaic_dataloader(
+            args.data_root,
+            batch_size=args.batch_size,
+            num_physical_nodes=num_physical_nodes,
+            epoch_size=target_epoch_size,
+            num_workers=torch_num_workers,
+        )
+
+    world_size = ray.train.get_context().get_world_size()
+    all_workers_time_list_across_epochs = []
     for i in range(args.num_epochs):
         print(f"Epoch {i+1} of {args.num_epochs}")
         num_rows = 0
         start_t = time.time()
-        for batch in it.iter_torch_batches(
-            batch_size=args.batch_size,
-        ):
-            num_rows += args.batch_size
-        end_t = time.time()
 
-    # Workaround to report the final epoch time from each worker, so that we
-    # can sum up the times at the end when calculating throughput.
-    world_size = ray.train.get_context().get_world_size()
-    all_workers_time_list = [
-        torch.zeros((2), dtype=torch.double, device=device) for _ in range(world_size)
-    ]
-    curr_worker_time = torch.tensor([start_t, end_t], dtype=torch.double, device=device)
-    dist.all_gather(all_workers_time_list, curr_worker_time)
+        # Ray Data needs to call iter_torch_batches on each epoch.
+        if isinstance(it, ray.data.iterator.DataIterator):
+            batch_iter = it.iter_torch_batches(
+                batch_size=args.batch_size,
+            )
 
+        print_at_interval = 1000
+        print_at = print_at_interval
+        for batch in batch_iter:
+            # `batch` should have tensor in `torch.Tensor` format.
+            if args.use_torch:
+                num_rows += batch.size(dim=0)
```

Comment:
Can we share this code for all versions? It'd be good to make sure that they all output the same format.

For ray, might need to pass the feature column.